Added code for rugby example

sudip-padhye opened the pull request:
1. I have added the rugby example code and the input CSV file.
2. The code to save the model has been added however output .nc files are ignored using .gitignore.
3. Plots are still present in the notebook as I was not sure whether to include them or not.

EDIT (Oriol): closes  arviz-devs/arviz#1294

sudip-padhye committed:
Added code for rugby example

canyon289 commented:
Havent looked into the notebook yet but its titled radon.ipynb?

sudip-padhye commented:
The notebook is titled radon in reference to the issue request (#1294) related to this Pull request. In that Issue request, we are asked to generate code for the radon dataset.

OriolAbril commented:
@ahartikainen I think we can reopen this one and close the other PR, this one is closer to be merged as it has already solved the merge issues. 

@sudip-padhye I have seen that while the title said rugby dataset in arviz-devs/arviz#1294, the body said radon, sorry about that. The notebook should be renamed. Moreover, if you look at the changes proposed by this PR, you'll see that is ways 4 files, currently, this PR is not only adding the rugby dataset but deleting the schools ones, you should undo the deletion.

sudip-padhye committed:
Undoing the schools data deletions and renaming radon to rugby

sudip-padhye commented:
@OriolAbril & @canyon289 Sorry for the confusion between the notebook names. I have renamed the filename from 'radon' to 'rugby'. Moreover, I have added the other two files 'centered_eight.ipynb' and 'non_centered_eight.ipynb' to the PR.

OriolAbril commented:
In addition to the contributing guide, this blog post may also help you get started with a git workflow: https://karinknudson.com/git_workflow.html

sudip-padhye commented:
@OriolAbril Thanks for sharing the guide. This will surely help me in getting more insights related to the git workflow. 

Moreover, I wanted to know are the changes that were pushed recently on par with the requirement? Or do I need to make any further changes?

OriolAbril commented:
There are still some changes to be done:

* If you look at the current inferencedata for rugby example you'll notice that the posterior_predictive and the prior groups are also present. The code should also add them to the inference data, it currently generates only a portion of the data. It should be a couple more lines to add to the code, the 8 schools example do have all the groups and should serve as a guide.
* Similarly, you'll see that the dimensions of the generated example and the dimensions in the current arviz example are different too, you should be able to fix this with the `coords` and `dims` arguments of `from_pymc3`, again, the 8 schools examples should serve as a guide. Alternatively, you can also opt for the brand new coords and dims within PyMC3 model itself; in that case, a good example is [this notebook](https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/multilevel_modeling.ipynb)
* I noticed there is a dataset added as a csv file but still there is a cell with pandas commands, merging and further preprocessing of the data. I would remove the `try` and force reading the csv file you are adding to the repository, and I would make it the completely preprocessed data, so that it can be read and used straight away. This way the focus is completely on the model and inferencedata generation
* Like in the other PR, I would remove the plot, here you already start to see what I was mentioning before about the plots not being necessarily useful/representative for all datasets. Leaving plots aside should make all examples more homogeneous.

sudip-padhye commented:
With reference to the point-1, when I try to add the statement:

`prior = pm.sample_prior_predictive()`

it is throwing an error **"ValueError: Cannot sample from Flat distribution."**

Can you provide any suggestions on how can I add prior groups to the inference data without generating errors?

Thank You!

OriolAbril commented:
You can use `pm.Normal(..., 0, 5)` instead for example, it should be wide enough to not modify the results while being a proper prior

sudip-padhye commented:
By declaring the prior using pm.Normal as below,

`prior = pm.Normal("prior", mu = 0, sd = 5)`

I encountered an error **'FreeRV' object has no attribute 'keys'** while generating arviz data object using arviz.from_pymc3.

Am I missing something in the pm.Normal()?

OriolAbril commented:
Yes, sorry. I'll try to be more clear and detailed now. 

PyMC3 is being used to define a Bayesian model that can then used for several tasks such as obtaining the probability distribution of the model parameters conditioned to the observed data (that would be `pm.sample`). 

To build this model, several elements (probability distributions) are combined. However, not all building blocks are compatible with all the tasks. Your original error was:

    ValueError: Cannot sample from Flat distribution.

which was triggered by `pm.sample_prior_predictive()`. This error is due to `pm.Flat` not being compatible with sampling the prior. Therefore, to be able to sample the prior, the `Flat` blocks have to be modified, for example, changing them for `Normal` which is compatible with prior sampling.

When defining `prior` as `pm.Normal` what is happening is that `az.from_pymc3` is expecting a python `dict` but it is instead receiving a `FreeRV` which is the type of these PyMC3 building blocks. As this `FreeRV` objects do not have the same attributes and properties as dicts, the program crashes with 

    'FreeRV' object has no attribute 'keys'

this means that ArviZ is trying to iterate over the keys of a dictionary but it does not have a dict, it has a FreeRV.

For more details, see https://docs.pymc.io/

sudip-padhye committed:
Added prior groups, coords and dims

sudip-padhye commented:
With reference to the above comments, here I push the following changes:

1] prior groups have been added to the model. Since, `prior = pm.Normal("prior", mu = 0, sd = 5)` is a FreeRV, first I have created trace containing the prior and later creating a dictionary containing "prior", "home_points" and "away_points". This eliminates the FreeRV error and successfully incorporates prior to the model.

2] The coords and dims have been using the no. of teams present in the rugby dataset, which is in analogy with the 8 schools example in the centered_eight dataset. 

3] All the input data preprocessing is done and saved as a CSV file. Thus, we no longer need to preprocess the data. Now, we can just read the modified CSV file and readily use it in the model creation.

4] Plots have been removed from the notebook.

OriolAbril commented:
Thanks @sudip-padhye ! 

I have added the next review iteration via reviewNB which refers to different cells. At the beggining of the PR there is a link to go to ReviewNB and see the discussion there, here part of the information is lost. If something is not clear enough do not hesitate to ask here or in [gitter](https://gitter.im/arviz-devs/community).

OriolAbril commented on rugby.ipynb:
```diff
@@ -0,0 +1,3579 @@
+{
```
<p>I think the definition of <code>teams</code> should be here. Moreover, some care is in order with its values. Now for example, you will see that while the first position of <code>teams</code> is England, however, according to the csv, the <code>0</code> index corresponds to Wales! :warning:</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz_example_data/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

sudip-padhye replied:
<p>The order of the values in teams is done with respect to their wins in each match. What criteria do you suggest for team ordering? Could you please provide the team ordering of the 6 teams?</p>

sudip-padhye replied:
<p>I just read the output of the posterior group for the example in arviz. As per the example the team order is <code>['Wales', 'France', 'Ireland', 'Scotland', 'Italy' 'England']</code> . Is it correct?</p>

OriolAbril replied:
Yes, exactly!

OriolAbril commented on rugby.ipynb:
```diff
@@ -0,0 +1,3579 @@
+{
```
<p>prior  should not be an extra variable here. To sample from the prior, sample_prior_predictive  must be used, following the same pattern as in the eight schools examples. This automatically generates a dictionary with all the variables in the model: home, sd_att, sd_def, ..., home_points, away_points that is used to populate prior and prior predictive groups (ArviZ knows how to split the dict into the two groups). As I commented before, <code>pm.Flat</code>  is incompatible with prior sampling. Therefore, all instances of it have to be removed and replaced by something compatible with prior sampling. This means modifiying two lines and deleting the <code>prior = pm.Norma...</code>  one:</p><p> </p><pre class="ql-syntax" spellcheck="false">home = pm.Normal('home', mu=0, sigma=5)
intercept = pm.Normal('intercept', mu=0, sigma=5)
</pre><p> </p><p>More on this below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz_example_data/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

sudip-padhye replied:
<p>Thank you for the detailed explanation. I have made the required changes in the file.</p>

OriolAbril commented on rugby.ipynb:
```diff
@@ -0,0 +1,3579 @@
+{
```
<p>This cell should be:</p><p> </p><pre class="ql-syntax" spellcheck="false">with model:
&nbsp;&nbsp;&nbsp; prior = pm.sample_prior_predictive()
&nbsp;&nbsp;&nbsp; trace = pm.sample(500, tune=500, cores=4)
&nbsp;&nbsp;&nbsp; posterior_predictive = pm.sample_posterior_predictive(trace)
</pre><p> </p><p>I'll try to explain the reason behind each change. </p><p> </p><p>Continuing on the comment above, to sample from the prior and thus get the data that goes into prior and prior predictive groups, <code>sample_prior_predictive</code>  has to be used, sadly the current code does not generate the variables for the prior and prior predictive groups.</p><p> </p><p>The current rugby example has 4 chains and 500 draws whereas the current code generates 1000 draws and 3 chains. The changes above should fix that difference</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz_example_data/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

sudip-padhye replied:
<p>I made the above changes in the cell however while executing the code, I am facing a new error <code>ValueError: lam value too large</code> for the code line <code>prior = pm.sample_prior_predictive()</code> .</p><p> </p><p>On searching for the error on google, I found that the error is due to certain numpy libraries used in pm.sample_prior_predictive() method. Moreover, this type of error is faced on Windows operating system. </p>

OriolAbril replied:
Can you try:

```
home = pm.Normal('home', mu=0, sigma=1)
intercept = pm.Normal('intercept', mu=3, sigma=1)
```

This should be more restrictive and fix the overflow issue you mentioned above

sudip-padhye replied:
<p>Unfortunately, even these changes are not working out. Program is encountering the same error. Is there any other workaround? </p>

OriolAbril replied:
Maybe with

```
sd_att = pm.HalfStudentT('sd_att', nu=3, sigma=1.5)
sd_def = pm.HalfStudentT('sd_def', nu=3, sigma=1.5)
```

or if this does not work either, with 

```
sd_att = pm.HalfNormal('sd_att', sigma=2)
sd_def = pm.HalfNormal('sd_def', sigma=2)
```

After this, push the changes independently of the code working or not. It it were still not working I'd play around to see what can be happening.

sudip-padhye replied:
I tried with the above code lines and the code 

`sd_att = pm.HalfNormal('sd_att', sigma=2)
sd_def = pm.HalfNormal('sd_def', sigma=2)` 

worked perfectly fine. The code prior to these lines still had an error. However, now I am facing another issue with the `arviz.from_pymc3(...., dims = {...})`. The code `dims={"atts": [teams], "defs": [teams], "atts_star": [teams], "defs_star": [teams]}` is throwing an error `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`. However, the teams are not boolean arrays but still the code is treating it as a boolean array.

OriolAbril replied:
Oh, my bad, `dims` is a mapping from variable names to dimension names, whereas now it is kind of a mapping between variable names and coordinate values. The correct syntax should be:

    dims={"atts": ["team"], "defs": ["team"], "atts_star": ["team"], "defs_star": ["team"]}

Note that the dimension name is `team` which is the key in `coords` (coords sets the values of `team` dimension to the values in array `teams`). Also, the variables whose dimension is `match` (home and away points, check the observed data group of the current example) should also be added to `dims`. I hope the pattern is clear now, sorry about the previous mix-up

sudip-padhye replied:
I modified the dims part as below:

`dims = {"atts": ["team"], "defs": ["team"], "atts_star": ["team"],  "defs_star": ["team"], "home_points": ["match"], "away_points": ["match"]}`

However, the ["team"] value is posing an error `ValueError: dimensions ('team',) must have the same length as the number of data dimensions, ndim=2`. Am I missing something in the modified `dims` definition?

OriolAbril replied:
Can you push the current code?

OriolAbril commented on rugby.ipynb:
```diff
@@ -0,0 +1,3579 @@
+{
```
<p>As you can see when exploring the dimensions and coordinates of the generated object, they are not the same as the example available from within ArviZ. If you explore all the groups, you will see that only 4 dimensions are present: chain, draw, team and match. In the idata generated here, several more are included. The usage of coords and dims in ArviZ is the following:</p><p> </p><ul><li>coords is a dictionary whose name are dimension names (thus team and match will be keys here, chain and draw are added by ArviZ so don't need to be here) and whose values are coordinate values. The dict you currently have is already correct regarding the team dimension, only the match one is missing.</li><li>dims is a dictionary whose values are variable names and whose values are lists of the dimensions for each variable. Here, the keys are theta and obs. theta and obs where variable names in the 8 schools model, but they are not here. Therefore, they are ignored and the team dimension is not added anywhere. The variables that should be added are the ones who have either team/match dimension in the arviz idata example. </li></ul><p> </p><p>Here is the print of the posterior group for the example in arviz:</p><p> </p><pre class="ql-syntax" spellcheck="false">&lt;xarray.Dataset&gt;
Dimensions:&nbsp;&nbsp;&nbsp; (chain: 4, draw: 500, team: 6)
Coordinates:
&nbsp; * chain&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (chain) int64 0 1 2 3
&nbsp; * draw&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (draw) int64 0 1 2 3 4 5 6 7 ... 492 493 494 495 496 497 498 499
&nbsp; * team&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (team) object 'Wales' 'France' 'Ireland' ... 'Italy' 'England'
Data variables:
&nbsp;&nbsp;&nbsp; home&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (chain, draw) float64 ...
&nbsp;&nbsp;&nbsp; intercept&nbsp; (chain, draw) float64 ...
&nbsp;&nbsp;&nbsp; atts_star&nbsp; (chain, draw, team) float64 ...
&nbsp;&nbsp;&nbsp; defs_star&nbsp; (chain, draw, team) float64 ...
&nbsp;&nbsp;&nbsp; sd_att&nbsp;&nbsp;&nbsp;&nbsp; (chain, draw) float64 ...
&nbsp;&nbsp;&nbsp; sd_def&nbsp;&nbsp;&nbsp;&nbsp; (chain, draw) float64 ...
&nbsp;&nbsp;&nbsp; atts&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (chain, draw, team) float64 ...
&nbsp;&nbsp;&nbsp; defs&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (chain, draw, team) float64 ...
</pre><p> </p><p>As you can see, atts and defs (as well as their star versions) have <code>team</code>  as dimension, thus, these variables should be keys in the <code>dims</code>  dict</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz_example_data/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

sudip-padhye replied:
<ol><li><strong>coords</strong>: I am little confused with what value must be mapped to the key "match"? Could you please help me in identifying the same.</li><li><strong>dims</strong>: I have modified the dims argument. Following is the new code:</li></ol><p>							<code>dims={"atts": teams, "defs": teams, "atts_star": teams, "defs_star": teams}</code>  </p><p> </p><p>Is this as per the requirement for the dims?</p><p> </p><p>Thank You!</p>

OriolAbril replied:
1. To generate the same coordinate values for match as the ones in the current ArviZ rugby example you can do the following:

```
match = [f"{home} {away}" for home, away in zip(df.home_team, df.away_team)]
```

2. Note that `dims` has to be a dictionary of lists, thus, it should be `dims={"atts": [teams], "defs": [teams], "atts_star": [teams], "defs_star": [teams]}`

sudip-padhye replied:
<p>Thank You @OriolAbril for your help and nice explanation!</p><p> </p><p>I have made the required changes for this section of the code. </p>

sudip-padhye committed:
Changed team ordering and added sample_prior_predictive

OriolAbril commented on rugby.ipynb:
```diff
@@ -0,0 +1,186 @@
+{
```
<p>I think I see the issue, <code>teams</code>  and <code>matches</code>  are already arrays, so they don't have to be putted inside a list: <code>coords = {"team": teams, "match": matches}</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/arviz_example_data/pull/4/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

sudip-padhye replied:
Oh, I see! That means python was trying to create an array of an array and hence the error. Now, I got it. Thanks for figuring that out! I will push the final code.

sudip-padhye committed:
Updated coords definition in arviz.from_pymc3

OriolAbril reviewed (approved):
Thanks for the PR!
